### DISCUSSION TITLE
"must_not" doesn't work?

### DISCUSSION THREAD

**Discussion by wkt** [milvus-io/milvus#4185] [Q&A and General discussion] [answered]
Code:
from milvus import Milvus, DataType

collection_name = 'test'
client = Milvus(host='127.0.0.1', port='19530')

collection_param = {
    "fields": [
        {"name": "c0", "type": DataType.INT64},
        {"name": "c1", "type": DataType.INT64},
        {"name": "vec", "type": DataType.FLOAT_VECTOR, "params": {"dim": 2}}
    ],
    "auto_id": True
}
ivf_param = {"index_type": "FLAT", "metric_type": "L2", "params": {"nlist": 2048}}

client.drop_collection(collection_name)
client.create_collection(collection_name, collection_param)
client.create_index(collection_name, "vec", ivf_param)

c0_list = [1, 2, 3, 4, 5, 6]
c1_list = [3, 4, 5, 6, 7, 8]
vec_list = [[0., 0.1], [0., 0.2], [0., 0.3], [0., 0.4], [0., 0.5], [0., 0.6]]

entities = [
    {"name": "c0", "values": c0_list, "type": DataType.INT64},
    {"name": "c1", "values": c1_list, "type": DataType.INT64},
    {"name": "vec", "values": vec_list, "type": DataType.FLOAT_VECTOR}
]

eids = client.insert(collection_name, entities)
client.flush()
ens = client.get_entity_by_id(collection_name, ids=[eids[0]], fields=['vec'])

vecs = [[0., 0.02], [0., 0.35]]
dsl = {

    "bool": {
        "must": [
            {
                "vector": {
                    "vec": {
                        "topk": 3,
                        "metric_type": "L2",
                        "query": vecs,
                        "params": {
                            "nprobe": 10
                        }
                    }
                },
            }
        ],
    },
    "must_not": [
        {
            "term": {
                "c0": [1, 2],
                "c1": [3, 6]
            }
        }
    ]
}

res = client.search(collection_name, dsl, fields=['c0', 'c1', 'vec'])
for y in res:
    for x in y:
        print(x.entity.get('c0'), x.entity.get('c1'), x.entity.get('vec'), "\n")

结果：
1 3 [0.0, 0.10000000149011612] 

2 4 [0.0, 0.20000000298023224] 

3 5 [0.0, 0.30000001192092896] 

3 5 [0.0, 0.30000001192092896] 

4 6 [0.0, 0.4000000059604645] 

2 4 [0.0, 0.20000000298023224] 

Data of c0 == 1,c0 == 2 and c1 == 6 were not filtered out.

**Comment by shiyu22** [comment, marked as answer]
You can refer to this:
```json
{
  "bool": {
    "must": [
      {
        "must": [
          {
            "vector": {
              "embedding": {
                "metric_type": "L2", 
                "query": [
                  [
                    -0.0882435888051987, ... 0.0524853430688381
                  ]
                ], 
                "topk": 100
              }
            }
          }
        ]
      }, 
      {
        "must_not": [
          {
            "term": {
              "customer_id": [
                -2020600124
              ]
            }
          }
        ]
      }
    ]
  }
}
```

**Reply by wkt** [reply]
Think your reply.
but it still not working.
It get an excpetion:
milvus.client.exceptions.BaseException: <BaseException: (code=1, message=The direct child node of 'bool' node cannot be 'should' node or 'must_not' node.)>

**Reply by fishpenguin** [reply]
There is a bug in must_not query in 0.11.0, we have already fixed it in 0.11.1 in [issue-4170](https://github.com/milvus-io/milvus/issues/4170)

**Reply by fishpenguin** [reply]
In 0.11.0, we only recommend you to use must in your dsl. Because there are some bugs in the validation check of 'should' and 'must_not' query, and these bugs will be fixed in 0.11.1.

**Reply by wkt** [reply]
> In 0.11.0, we only recommend you to use must in your dsl. Because there are some bugs in the validation check of 'should' and 'must_not' query, and these bugs will be fixed in 0.11.1.

It milvus-0.10.3 has this bug?

**Reply by sahuang** [reply]
Milvus 0.10.3 does not support DSL query.

**Reply by wkt** [reply]
> Milvus 0.10.3 does not support DSL query.

OK,when 0.11.1 will be release ?

**Reply by sahuang** [reply]
There isn't a precise timeline but it will be released after all bugs are cleared.
